pyquil/operator_estimation.py

```diff
@@ -300,10 +302,51 @@ def _ops_diagonal_in_tpb(op_code1: str, op_code2: str):
 def _all_qubits_diagonal_in_tpb(op1: PauliTerm, op2: PauliTerm):
     """
     Compare all qubits between two PauliTerms to see if they are all diagonal in an
-    overall shared tensor product basis.
+    overall shared tensor product basis. More concretely, test if ``op1`` and
+    ``op2`` are diagonal in each others' "natural" tensor product basis.
+
+    Given some PauliTerm, the 'natural' tensor product basis (tpb) to
+    diagonalize this term is the one which diagonalizes each Pauli operator in the
+    product term-by-term.
+
+    For example, X(1) * Z(0) would be diagonal in the 'natural' tensor product basis
+    {(|0> +/- |1>)/Sqrt[2]} * {|0>, |1>}, whereas Z(1) * X(0) would be diagonal
+    in the 'natural' tpb {|0>, |1>} * {(|0> +/- |1>)/Sqrt[2]}. The two operators
+    commute but are not diagonal in each others 'natural' tpb (in fact, they are
+    anti-diagonal in each others 'natural' tpb). This function tests whether two
+    operators given as PauliTerms are both diagonal in each others 'natural' tpb.
+
+    Note that for the given example of X(1) * Z(0) and Z(1) * X(0), we can construct
+    the following basis which simultaneously diagonalizes both operators:
+
+      -- 2 * |0>' = |0> (|0> + |1>) + |1> (|0> - |1>)
+      -- 2 * |1>' = |0> (|0> + |1>) - |1> (|0> - |1>)
+      -- 2 * |2>' = |0> (|0> - |1>) + |1> (|0> + |1>)
+      -- 2 * |3>' = |0> (-|0> + |1>) + |1> (|0> + |1>)
+
+    In this basis, X Z looks like diag(1, -1, 1, -1), and Z X looks like diag(1, 1, -1, -1)
+
+    :param op1: PauliTerm to check diagonality of in the natural tpb of ``op2``
+    :param op2: PauliTerm to check diagonality of in the natural tpb of ``op1``
+    :return: Boolean of diagonality in each others natural tpb
     """
-    all_qubits = set(op1.get_qubits()) | set(op2.get_qubits())
-    return all(_ops_diagonal_in_tpb(op1[q], op2[q]) for q in all_qubits)
+    all_qubits = set(op1.get_qubits()) & set(op2.get_qubits())
```

Comment:
why did you change this from `|` to `&`?

Reply:
I think I mentioned this in another comment here somewhere. No special reason except that there was another function which did basically the same thing except it had `&` instead of `|`, which seemed safer to me against any possible future changes to `PauliTerm` (e.g. if I have `pt = (1+0j)*Z0*X1*Y2*Z5` and then try to get `pt[3]`, in case any future change returns some error instead of returning the identity, which it currently does). I deprecated that other function, but retained this simple change.